src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/GaugeWave.cpp

```diff
@@ -43,293 +40,320 @@ DataType gauge_wave_deriv_h(const tnsr::I<DataType, 3>& x, const double t,
 
 namespace gr {
 namespace Solutions {
-GaugeWave::GaugeWave(const double amplitude, const double wavelength) noexcept
+template <size_t Dim>
+GaugeWave<Dim>::GaugeWave(const double amplitude,
+                          const double wavelength) noexcept
     : amplitude_(amplitude), wavelength_(wavelength) {
   ASSERT(amplitude > 0.0,
          "Amplitude must be non-negative. Given amplitude: " << amplitude_);
   ASSERT(wavelength > 0.0,
          "Wavelength must be non-negative. Given wavelength: " << wavelength_);
 }
 
-void GaugeWave::pup(PUP::er& p) noexcept {
+template <size_t Dim>
+void GaugeWave<Dim>::pup(PUP::er& p) noexcept {
   p | amplitude_;
   p | wavelength_;
 }
 
+template <size_t Dim>
 template <typename DataType>
-GaugeWave::IntermediateVars<DataType>::IntermediateVars(
+GaugeWave<Dim>::IntermediateVars<DataType>::IntermediateVars(
     const double amplitude, const double wavelength,
-    const tnsr::I<DataType, GaugeWave::volume_dim, Frame::Inertial>& x,
+    const tnsr::I<DataType, volume_dim, Frame::Inertial>& x,
     const double t) noexcept {
   h = gauge_wave_h(x, t, amplitude, wavelength);
   dx_h = gauge_wave_deriv_h(x, t, amplitude, wavelength);
   sqrt_h = sqrt(h);
   dx_h_over_2_sqrt_h = 0.5 * dx_h / sqrt(h);
 }
 
+template <size_t Dim>
 template <typename DataType>
-auto GaugeWave::variables(
-    const tnsr::I<DataType, GaugeWave::volume_dim, Frame::Inertial>& /*x*/,
-    const double /*t*/, const GaugeWave::IntermediateVars<DataType>& vars,
+auto GaugeWave<Dim>::variables(
+    const tnsr::I<DataType, volume_dim, Frame::Inertial>& /*x*/,
+    const double /*t*/, const IntermediateVars<DataType>& vars,
     tmpl::list<gr::Tags::Lapse<DataType>> /*meta*/) const noexcept
     -> tuples::TaggedTuple<gr::Tags::Lapse<DataType>> {
   return {Scalar<DataType>{vars.sqrt_h}};
 }
 
+template <size_t Dim>
 template <typename DataType>
-auto GaugeWave::variables(
-    const tnsr::I<DataType, GaugeWave::volume_dim, Frame::Inertial>& /*x*/,
-    const double /*t*/, const GaugeWave::IntermediateVars<DataType>& vars,
+auto GaugeWave<Dim>::variables(
+    const tnsr::I<DataType, volume_dim, Frame::Inertial>& /*x*/,
+    const double /*t*/, const IntermediateVars<DataType>& vars,
     tmpl::list<::Tags::dt<gr::Tags::Lapse<DataType>>> /*meta*/) const noexcept
     -> tuples::TaggedTuple<::Tags::dt<gr::Tags::Lapse<DataType>>> {
   return {Scalar<DataType>{-1.0 * vars.dx_h_over_2_sqrt_h}};
 }
 
+template <size_t Dim>
```

Comment:
Why not `template <size_t Dim, typename DataType>`?

Reply:
The first one is the templates for the class and the second is for the function.  It is slightly odd syntax, but that's how you do member function templates of class templates.